### ISSUE TITLE
Download speed limit is not obeyed

### ISSUE THREAD

**Issue by nengxu** [transmission/transmission#7005]
### What is the issue?

This issue has existed for a while. The last version I tried which did obey the speed limit is transmission-4.1.0-dev+ree4daa059d-x64.

1) The global Download Speed Limit in Preference  | Speed is not effective. 

2) The per torrent Download Speed Limit in Torrent Propertios | Options is not effective.

3) No matter which values other than 0 assigned to the above 2 limits, the actual download speed for one torrent seems to be around 8.19KB/s for each peer connected. Usually there will be 4 concurrent peers connected for downloading simultaneously. So the total download speed for one torrent would be around 32.77KB/s. 

4) Only when the Limit download speed for one torrent is set to 0, the limit is effective. But this is useless as no downloading is happening.

5) If Maximum peers is set to 1, the total speed for one torrent is still 32.77KB/s, even though only one peer is connected. 

6) If Maximum peers is set to moren than 1, the total speed for one torrent is still 32.77KB/s, though more peers are connected for downloading. 

7) No matter Maximum peers is set to which number bigger than 4, the concurrently connected peers seem to be 4 at most of the time.

So it seems that there is a virtual minimum limit of 32.77KB/s for one torrent, and a per peer minimum limit of 8.19KB/s, no matther how those Speed limits are set.

8) I retested on the transmission-4.1.0-dev+ree4daa059d-x64 which obeyed the limits. No matter which value assigned to per torrent Download Speed Limit (eg, 5 or 4), the minimum speed per peer is 8KB/s. However, it take too long to acuatually download form a peer at the speed of 8KB/s. In other words, the downloading happends infrequently. But the total speed for a torrent can more or less obey the limit. In my case, I set the limit to be 8KB/s. the the total speed is either 8KB/s (one connected peer) or 16KB/s (2 connected peers), or sometimes 24KB (3 connected peers). Though not 32KB/s as in newer version of Transmission.

9)  The global Download Speed Limit is obeyed in transmission-4.1.0-dev+ree4daa059d-x64.

### Which application of Transmission?

Qt app on Windows

### Which version of Transmission?

transmission-4.1.0-dev+r3e9f5f614a-x64

**Comment by killemov** [comment]
Duplicate of #6736. Nice analysis though.

**Comment by nengxu** [comment]
I read through #6736 . Though the problem is still happening with every new night build I tried. Please fix the problem. And until the problem has been fixed in a new release, please don't close this issue like someone did with #6736. Thanks.

> Duplicate of #6736. Nice analysis though.

**Comment by killemov** [comment]
@tearfur Somebody spoke up?

**Comment by tearfur** [comment]
@nengxu Thanks for reporting. I'm currently focused on issue 6986, I will get around to this when I have the time, or hopefully somebody else can do it before I do.

***

@killemov Did you spot where the bug is in the code? I don't know how you can be so sure that this is a duplicate of 6736. It feels like a deja vu, I've asked you before to bring out more concrete evidences if you have an idea of what's going on, so that everybody else can catch on faster.

Or, if this is just a tactic of yours to provoke me (or other contributors) to respond faster, then I hate it, please stop.

**Comment by killemov** [comment]
I don't go digging in the c++ code. The symptoms a similar (the same) and since you closed the other issue I thought of mentioning you. No ulterior motives here. So maybe "related to" instead of "duplicate of" is the better wording here?

**Comment by tearfur** [comment]
Just tested on 3e5a77d17603dd6548add3c5b233a4e87d38c577 and cannot reproduce.

![test](https://github.com/user-attachments/assets/3b93ad6f-e421-47ec-99e3-9f03dcd67d22)

The binary was built by myself, and every setting was on default.

```
cmake -G Ninja -B build_reldeb -DCMAKE_BUILD_TYPE=RelWithDebInfo -DENABLE_TESTS=NO -DRUN_CLANG_TIDY=NO -DCMAKE_TOOLCHAIN_FILE=C:\vcpkg\scripts\buildsystems\vcpkg.cmake
```

**Comment by nengxu** [comment]
> Just tested on [3e5a77d](https://github.com/transmission/transmission/commit/3e5a77d17603dd6548add3c5b233a4e87d38c577) and cannot reproduce.
> 
> ![test](https://private-user-images.githubusercontent.com/46261767/360947847-3b93ad6f-e421-47ec-99e3-9f03dcd67d22.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ0MjQ1MjMsIm5iZiI6MTcyNDQyNDIyMywicGF0aCI6Ii80NjI2MTc2Ny8zNjA5NDc4NDctM2I5M2FkNmYtZTQyMS00N2VjLTk5ZTMtOWYwM2RjZDY3ZDIyLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIzVDE0NDM0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBkOGY4MzY3MWE0NmQ0N2E3NmI3ZTAwNWY3YzY2NWNlYTA1ODE4ODczOGRiOWE2ZGEzYzY1NGYxNmNlMTg4YzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.XT0wT6DC3zcmbDLYRyM6Nh3SYzmZUlZ8xOOpS9V98xg) [ ![test](https://private-user-images.githubusercontent.com/46261767/360947847-3b93ad6f-e421-47ec-99e3-9f03dcd67d22.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ0MjQ1MjMsIm5iZiI6MTcyNDQyNDIyMywicGF0aCI6Ii80NjI2MTc2Ny8zNjA5NDc4NDctM2I5M2FkNmYtZTQyMS00N2VjLTk5ZTMtOWYwM2RjZDY3ZDIyLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIzVDE0NDM0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBkOGY4MzY3MWE0NmQ0N2E3NmI3ZTAwNWY3YzY2NWNlYTA1ODE4ODczOGRiOWE2ZGEzYzY1NGYxNmNlMTg4YzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.XT0wT6DC3zcmbDLYRyM6Nh3SYzmZUlZ8xOOpS9V98xg) ](https://private-user-images.githubusercontent.com/46261767/360947847-3b93ad6f-e421-47ec-99e3-9f03dcd67d22.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ0MjQ1MjMsIm5iZiI6MTcyNDQyNDIyMywicGF0aCI6Ii80NjI2MTc2Ny8zNjA5NDc4NDctM2I5M2FkNmYtZTQyMS00N2VjLTk5ZTMtOWYwM2RjZDY3ZDIyLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIzVDE0NDM0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBkOGY4MzY3MWE0NmQ0N2E3NmI3ZTAwNWY3YzY2NWNlYTA1ODE4ODczOGRiOWE2ZGEzYzY1NGYxNmNlMTg4YzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.XT0wT6DC3zcmbDLYRyM6Nh3SYzmZUlZ8xOOpS9V98xg) [ ](https://private-user-images.githubusercontent.com/46261767/360947847-3b93ad6f-e421-47ec-99e3-9f03dcd67d22.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ0MjQ1MjMsIm5iZiI6MTcyNDQyNDIyMywicGF0aCI6Ii80NjI2MTc2Ny8zNjA5NDc4NDctM2I5M2FkNmYtZTQyMS00N2VjLTk5ZTMtOWYwM2RjZDY3ZDIyLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIzVDE0NDM0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBkOGY4MzY3MWE0NmQ0N2E3NmI3ZTAwNWY3YzY2NWNlYTA1ODE4ODczOGRiOWE2ZGEzYzY1NGYxNmNlMTg4YzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.XT0wT6DC3zcmbDLYRyM6Nh3SYzmZUlZ8xOOpS9V98xg)
> 
> The binary was built by myself, and every setting was on default.
> 
> ```
> cmake -G Ninja -B build_reldeb -DCMAKE_BUILD_TYPE=RelWithDebInfo -DENABLE_TESTS=NO -DRUN_CLANG_TIDY=NO -DCMAKE_TOOLCHAIN_FILE=C:\vcpkg\scripts\buildsystems\vcpkg.cmake
> ```

Great! When can this fix be in a public build?

**Comment by tearfur** [comment]
It indeed does not go below 32KB most of the time, and I think I know why that is.

Bad news is, it has to do with the fundamental program design, so it is not so easy to fix...

**Comment by nengxu** [comment]
> It indeed does not go below 32KB most of the time, and I think I know why that is.
> 
> Bad news is, it has to do with the fundamental program design, so it is not so easy to fix...

As I reported in the description, transmission-4.1.0-dev+ree4daa059d-x64 can maintain a low limit of 8KB for most of the time, though sometimes the speed can go up to 16KB, and rarely up to 24KB. Did the fundamental program design changed after the one I used?

**Comment by tearfur** [comment]
> transmission-4.1.0-dev+ree4daa059d-x64

Can you link me the commit ID? I'm not too sure where you got that string from.

**Comment by nengxu** [comment]
> > transmission-4.1.0-dev+ree4daa059d-x64
> 
> Can you link me the commit ID? I'm not too sure where you got that string from.

This string is in the name of the daily development build for Windows (transmission-4.1.0-dev+ree4daa059d-x64.msi). It is not associated with a commit. The date for this specific build is around 2023-10-01.

**Comment by selimachour** [comment]
I'm on version 4.0.5 (64805552b4) on Linux Mint 22 Wilma and still have the same problem.
I set alternative download limit to 200 kB/s, the turtle is clicked but I'm still downloading at over 1MB/s (full bandwidth)

**Comment by nengxu** [comment]
Any progress on this issue? Happy holidays!

**Comment by tearfur** [comment]
@selimachour

Speed limit and the turtle button on 4.0.5 is bugged. They should be fixed by #6551 and #6483+#6937 respectively.

@nengxu

The crux of the issue is that Transmission's speed limiter acts on the "piece speed", which is measuring the data flow of ***only*** BT piece data.

The limitation with that is BT pieces are generally transmitted in blocks of 16KiB, that means 16KiB is the smallest increment Transmission can work with when defining speed limits, hence why you were having problems with a 8kB/s limit. In more tech terms, the granularity of the current speed limiter implementation is 16KiB/s. (You may still see the speed drop below 16kB/s occassionally, but that's a whole other mess related to timings within Transmissions)

So far, the most obvious solution I see is to have the speed limiter act on the ***raw*** speed (protocol overhead included), which has smaller granularity than the piece speed.

**Comment by akane741** [comment]
8.19kB/s!  I am on 4.1.0 and I got this speed very frequently.  I searched this magic number in the search engine and got this.

**Comment by tearfur** [comment]
8.19kB/s is a bug, but the bug also masks a fault in the uTP bandwidth code that causes the download speed limiter to overshoot, and that is a bigger issue than this bug.

So the uTP part of the fix #7339 is reverted to mask the overshooting problem for the time being.

The uTP code is overhauled in #8060 to address the fault, and it's currently awaiting code review.
